src/coreclr/interpreter/compiler.cpp

```diff
@@ -1914,11 +2084,20 @@ int InterpCompiler::GenerateCode(CORINFO_METHOD_INFO* methodInfo)
         AddIns(INTOP_BREAKPOINT);
 #endif
 
-    if ((methodInfo->options & CORINFO_OPT_INIT_LOCALS) && m_ILLocalsSize > 0)
+    // We need to always generate this opcode because even if we have no IL locals, we may have
+    //  global vars which contain managed pointers or interior pointers
+    m_pInitLocalsIns = AddIns(INTOP_INITLOCALS);
+
+    if (methodInfo->options & CORINFO_OPT_INIT_LOCALS)
+    {
+        m_pInitLocalsIns->data[0] = m_ILLocalsOffset;
+        m_pInitLocalsIns->data[1] = m_ILLocalsSize;
+    }
+    else
     {
-        AddIns(INTOP_INITLOCALS);
-        m_pLastNewIns->data[0] = m_ILLocalsOffset;
-        m_pLastNewIns->data[1] = m_ILLocalsSize;
+        // Generate a 0-byte memset at the *end* of IL locals so when we extend it later we won't zero the IL locals
+        m_pInitLocalsIns->data[0] = m_ILLocalsOffset + m_ILLocalsSize;
```

Comment:
This doesn't seem correct. Aren't slots for IL locals reported the same as the ones for the other global vars ? So they are untracked, meaning they are seen as containing a ref / or not for the entire lifetime of the method. So we can hit the same problem with a ref IL local containing uninitialized junk reported to the GC, meaning we always need to 0 them as well.

Reply:
It sounds like we can't actually optimize out initlocals then unless a method has no byrefs/gc pointers in its locals?

Reply:
Not at this stage. I would expect at some point to have precise tracking for the liveness of these global vars.